### PR TITLE
Add feature to aid `ceph-mon` migration

### PR DESCRIPTION
**Pull request by tj90241**
Add feature to assist migration of `ceph-mons`
    
This PR adds a feature to the Nova migration code to
overwrite `ceph-mon` addresses in the domain's XML
specification with those specified in the general section
of a `/etc/ceph/migration.conf` configuration file.
    
It does the same for newly created and refreshed Nova
block device mappings and Cinder attachments as well.
    
Together, this assists the IP address migration of `ceph-mon`
nodes, especially for VMs which have a long uptime, by
allowing an operator to reconfigure the set of `ceph-mon`
IPs via a single live-migration of each VM in the cluster.

### CODE REVIEW

**Comment by tj90241** [comment]
Diff to `migration.py` can be found below:
```
--- migration.py	2021-11-12 16:58:40.000000000 +0000
+++ migration.py	2022-07-29 14:52:12.478949215 +0000
@@ -18,6 +18,8 @@
 """
 
 from collections import deque
+from configparser import ConfigParser
+from ipaddress import AddressValueError, IPv4Address
 
 from lxml import etree
 from oslo_log import log as logging
@@ -86,6 +88,7 @@
     xml_doc = _update_graphics_xml(xml_doc, migrate_data)
     xml_doc = _update_serial_xml(xml_doc, migrate_data)
     xml_doc = _update_volume_xml(xml_doc, migrate_data, get_volume_config)
+    xml_doc = _rewrite_ceph_monitor_hosts(xml_doc)
     xml_doc = _update_perf_events_xml(xml_doc, migrate_data)
     xml_doc = _update_memory_backing_xml(xml_doc, migrate_data)
     if get_vif_config is not None:
@@ -276,6 +279,40 @@
     return xml_doc
 
 
+def _rewrite_ceph_monitor_hosts(xml_doc):
+    """Rewrite monitor addresses for RBD volumes in the domain configuration
+
+    Extract monitor addresses from ceph.conf and overwrite whatever monitor
+    addresses associated with RBD volumes for the domain with these ones.
+    """
+    mons = []
+
+    try:
+        with open('/etc/ceph/ceph.conf', 'r') as config_file:
+            config = ConfigParser()
+            config.read_file(config_file)
+            mons = [h.strip() for h in config['global']['mon host'].split(',')]
+            validated_mons = [IPv4Address(host) for host in mons]
+    except (FileNotFoundError, KeyError, ValueError) as exception:
+        LOG.error('ceph-mon-migration: {0}'.format(str(exception)))
+        return xml_doc
+
+    for disk in xml_doc.findall('./devices/disk/source'):
+        if disk.get('protocol') != 'rbd':
+            continue
+
+        # Wax all the existing host pairs...
+        for host in disk.findall('host'):
+            disk.remove(host)
+
+        # ...and slot in the new mons
+        for mon_host in mons:
+            host = etree.SubElement(disk, 'host')
+            host.set('name', mon_host)
+            host.set('port', '6789')
+    return xml_doc
+
+
 def _update_perf_events_xml(xml_doc, migrate_data):
     """Update XML by the supported events of destination host."""
```

We take values from `ceph.conf` and rewrite the domain specification during migration time. A debug line is included in the exception case if something goes wrong, and the domain specification is left unchanged if we cannot derive the new monitor addresses.

**Comment by tj90241** [comment]
Another patch to overwrite the hosts when refreshing connection info, which fixes `nova.block_device_mapping`s:
```
--- block_device.py	2021-11-12 16:58:40.000000000 +0000
+++ block_device.py	2022-07-29 16:44:32.135548079 +0000
@@ -12,7 +12,9 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+from configparser import ConfigParser
 import functools
+from ipaddress import AddressValueError, IPv4Address
 import itertools
 
 from os_brick import encryptors
@@ -699,6 +701,23 @@
             if multiattach:
                 connection_info['multiattach'] = True
 
+        # Rewrite the connection_info using the list of mons in ceph.conf
+        mons = []
+
+        try:
+            with open('/etc/ceph/ceph.conf', 'r') as config_file:
+                config = ConfigParser()
+                config.read_file(config_file)
+                mon_host = config['global']['mon host']
+                t_mons = [h.strip() for h in mon_host.split(',')]
+                validated_mons = [IPv4Address(host) for host in mons]
+                mons = t_mons
+        except (FileNotFoundError, KeyError, ValueError) as exception:
+            LOG.error('ceph-mon-migration: {0}'.format(str(exception)))
+
+        if len(mons) and 'hosts' in connection_info.get('data', {}):
+            connection_info['data']['hosts'] = mons
+
         if 'serial' not in connection_info:
             connection_info['serial'] = self.volume_id
         self._preserve_multipath_id(connection_info)
```

**Comment by tj90241** [comment]
And, finally, for cinder attachments:
```
--- rbd.py	2022-07-29 18:09:24.918691061 +0000
+++ rbd.py	2022-07-29 18:12:37.993553548 +0000
@@ -14,7 +14,9 @@
 """RADOS Block Device Driver"""
 
 import binascii
+from configparser import ConfigParser
 import errno
+from ipaddress import AddressValueError, IPv4Address
 import json
 import math
 import os
@@ -509,6 +511,17 @@
             host, port = host_port.rsplit(':', 1)
             hosts.append(host.strip('[]'))
             ports.append(port)
+        # Overwrite the mon addrs using those in ceph.conf, not the monmap
+        try:
+            with open('/etc/ceph/ceph.conf', 'r') as config_file:
+                config = ConfigParser()
+                config.read_file(config_file)
+                mon_host = config['global']['mon host']
+                t_mons = [h.strip() for h in mon_host.split(',')]
+                validated_mons = [IPv4Address(host) for host in mons]
+                hosts = t_mons
+        except (FileNotFoundError, KeyError, ValueError) as exception:
+            LOG.error('ceph-mon-migration: {0}'.format(str(exception)))
         return hosts, ports
 
     def _get_usage_info(self):
```

**Review by psipika — DISMISSED** [review]

File: chef/cookbooks/bcpc/files/default/nova/block_device.py [hunk 1]
```diff
@@ -0,0 +1,963 @@
...
+                config = ConfigParser()
+                config.read_file(config_file)
+                mon_host = config['global']['mon host']
+                t_mons = [h.strip() for h in mon_host.split(',')]
+                validated_mons = [IPv4Address(host) for host in t_mons]
+                mons = t_mons
+        except (FileNotFoundError, KeyError, ValueError) as exception:
+            LOG.error('ceph-mon-migration: {0}'.format(str(exception)))
+
+        if len(mons) and 'hosts' in connection_info.get('data', {}):
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/nova/block_device.py:718]: Nit - although it's equivalent in Python, I'd suggest instead

```
if len(mons) > 0 and 'hosts' in connection_info.get('data', {}):
```

**Review by comay — DISMISSED** [review]

**Review by comay — APPROVED** [review]

**Pull request merged**